src/inmanta/plugins.py

```diff
@@ -259,188 +391,268 @@ def __init__(self, namespace: Namespace) -> None:
+        self.return_type = PluginReturn(get_annotation("return"))
 
-        def process_arg(index: int, argument_name: str, is_kwonly_arg: bool) -> None:
-            if argument_name not in arg_spec.annotations:
-                raise CompilerException(f"All arguments of plugin '{function.__name__}' should be annotated")
-            annotation = arg_spec.annotations[argument_name]
-            if annotation == Context:
-                self._context = index
-            else:
-                if is_kwonly_arg:
-                    plugin_argument = process_kw_only_args(argument_name, annotation)
-                else:
-                    plugin_argument = process_regular_args(index, argument_name, annotation)
-                arguments.append(plugin_argument)
+        if arg_spec.varargs is not None:
+            # We have a catch-all positional arguments
+            self.var_args = PluginArgument(
+                arg_name=arg_spec.varargs,
+                arg_type=get_annotation(arg_spec.varargs),
+            )
 
-        # Process regular arguments
-        for i in range(len(arg_spec.args)):
-            process_arg(i, arg_spec.args[i], is_kwonly_arg=False)
-        # Process keyword-only arguments
-        for i in range(len(arg_spec.kwonlyargs)):
-            process_arg(i, arg_spec.kwonlyargs[i], is_kwonly_arg=True)
+        if arg_spec.varkw is not None:
+            # We have a catch-all keyword arguments
+            self.var_kwargs = PluginArgument(
+                arg_name=arg_spec.varkw,
+                arg_type=get_annotation(arg_spec.varkw),
+            )
 
-        if "return" in arg_spec.annotations:
-            self._return = arg_spec.annotations["return"]
+        # Save all positional arguments
+        defaults_start_at = len(arg_spec.args) - len(arg_spec.defaults or [])
+        for position, arg in enumerate(arg_spec.args):
+            annotation = get_annotation(arg)
+            if annotation == Context:
+                self._context = position
+                continue
+
+            if self._context != -1:
+                # If we have a context argument, the position index
+                # needs to be adapted as this context object can never be passed
+                # from the model.
+                position -= 1
+
+            self.args.append(
+                PluginArgument(
+                    arg_name=arg,
+                    arg_type=annotation,
+                    arg_position=position,
+                    default_value=(
+                        arg_spec.defaults[position - defaults_start_at]
+                        if position >= defaults_start_at
+                        else PluginArgument.NO_DEFAULT_VALUE_SET
+                    ),
+                )
+            )
 
-        return arguments
+            # All positional arguments can also be assigned as kwargs
+            self.kwargs[arg] = self.args[position]
```

Comment:
Slight preference to using a temp var for this object to be a bit more robust.

Reply:
I don't really get what you mean with temp var?  And which object are you referring to?

Reply:
```
some_var = <the_new_arg>
self.args.append(some_var)
self.kwargs[arg] = some_var
```

So that we don't have to rely on the assumption that we inject exactly one object into `self.args` for each iteration.